### PR TITLE
Add AllOf, OneOf, AnyOf, Not and Dict types 

### PR DESCRIPTION
**Pull request by pietgeursen**

### CODE REVIEW

File: some.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+var is = require('./is')
+
+module.exports = function some(types, value) {
+  return types.some((type) => is(type, value))
```
Comment (ahdinosaur) [inline some.js]: this isn't a type.

File: index.js [hunk 2]
```diff
@@ -4,7 +4,10 @@ module.exports = Object.assign(
     validate: require('./validate'),
     is: require('./is'),
     struct: require('./struct'),
-    compose: require('./compose')
+    compose: require('./compose'),
+    maybe: require('./maybe'),
+    some: require('./some'),
+    dict: require('./dictionary'),
```
Comment (ahdinosaur) [inline index.js]: i think these new additions should be "types", and maybe with capital letters. (at least `Dict` should be capital, but why not all of them).

File: some.js (hunk 1 above)
Comment (pietgeursen) [inline some.js]: Yeah good point. Fixed. Makes the code for `Maybe` nicer too.

File: index.js (hunk 2 above)
Comment (pietgeursen) [inline index.js]: Cool done.

File: OneOf.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+var is = require('./is')
+
+module.exports = (types) => (value) => 
+  types.filter((type) => is(type, value)).length === 1
+    ? value
+    : new TypeError(`Expected ${value} to be on of ${types}`)
```
Comment (ahdinosaur) [inline OneOf.js]: `s/on/one`

File: index.js [hunk 4]
```diff
@@ -4,10 +4,11 @@ module.exports = Object.assign(
     validate: require('./validate'),
     is: require('./is'),
     struct: require('./struct'),
-    compose: require('./compose'),
-    Maybe: require('./maybe'),
-    Any: require('./any'),
-    Dict: require('./dictionary')
+    Maybe: require('./Maybe'),
```
Comment (ahdinosaur) [inline index.js]: these should be exported from `./types.js`

**Comment by ahdinosaur** [comment]
squash and merge when you're ready. :whale:

**Pull request merged**